pyopencl/array.py

```diff
@@ -612,15 +612,11 @@ def __init__(
 
                 shape = (shape,)
 
-            size = 1
-            for dim in shape:
-                size *= dim
-                if dim < 0:
-                    raise ValueError(
-                        f"negative dimensions are not allowed: {shape}")
-
-            if isinstance(size, np.integer):
-                size = size.item()
+            # Previously the size was computed as "size = 1; size *= dim for dim in shape
+            # However it can fail when using data types, eg., numpy.uint64(1) * 2  returns 2.0 !
+            if np.any(np.array(shape) < 0):
+                raise ValueError(f"negative dimensions are not allowed: {shape}")
+            size = np.prod(np.uint64(shape)).item()
```

Comment:
Given the number of corner cases we discovered, I think it would be good to add some tests, so that we don't accidentally re-break things in the future.